container/state.go

```diff
@@ -292,6 +292,9 @@ func (s *State) SetStopped(exitStatus *ExitStatus) {
 	}
 	s.ExitCodeValue = exitStatus.ExitCode
 	s.OOMKilled = exitStatus.OOMKilled
+
+	s.saveWaitResult(s.waitStopResult)
+	s.waitStopResult = new(StateStatus)
```

Comment:
I'm trying to get my head around these;

- we're passing a reference to a property of `s` to a method of `s`
- update the passed reference
- then immediately overwrite it with a new value

```go
func (s *State) saveWaitResult(result *StateStatus) {
 	result.exitCode = s.ExitCodeValue
 	result.err = s.Err()
 }
```


So effectively that's equivalent to;

```go
func (s *State) saveWaitResult() {
	s.waitStopResult = &StateStatus{
		exitCode: s.ExitCodeValue,
		err:      s.Err(),
	}
}
```

Which would require two separate functions for `saveWaitStopResult()` and `saveWaitRemoveResult()`, unless we can use the same (see my earlier comment)


Even then, I'm somewhat confused by the line following it;

```go
s.waitStopResult = new(StateStatus)
```

Reply:
They are not equivalent.
`saveWaitResult` updates the value under reference, not the reference itself.
It's intentional - the State also holds the statuses as pointers. 
What `Wait` now does is stores a copy of these pointers which represent a StateStatus that is not yet filled, but will be filled once the corresponding wait channel is closed. 

```go
stopResult := s.waitStopResult
removeResult := s.waitRemoveResult
```

So `saveWaitResult` writes result to a StateStatus that `Wait` calls have pointer to. After that creates a fresh StateStatus that will be filled by the next exit/removal - this still doesn't affect waiting `Wait` calls, because they still have a pointer to the StateStatus that they are watching. 
This wouldn't be the case with your `saveWaitResult`, because it doesn't modify the StateStatus, it creates a new one.

I agree though that it's not really obvious in the code - maybe there is some cleaner way to express this?